### PR TITLE
Quirk processing refactor & optimizations

### PR DESCRIPTION
**Pull request by utilbird**
## About The Pull Request

Quirks now use parent qdel signals for destruction handling instead of checking if their parent is being qdeleted every single time SSquirks fires. Now only quirks that _need_ to process will process, passive quirks will have basically no overhead.

Also makes some misc optimizations and code qol improvements

## Changelog
:cl:
code: misc optimizations to quirk code
refactor: refactored quirk processing
/:cl:

### CODE REVIEW

File: code/datums/traits/_quirk.dm [hunk 1]
```diff
@@ -10,6 +10,7 @@
 	var/medical_record_text //This text will appear on medical records for the trait. Not yet implemented
 	var/mood_quirk = FALSE //if true, this quirk affects mood and is unavailable if moodlets are disabled
 	var/mob_trait //if applicable, apply and remove this mob trait
+	var/process = FALSE // Does this proc use on_process()?
```
Comment (utilbird) [inline code/datums/traits/_quirk.dm]: Cleaner than just putting START_PROCESSING in the New() of every quirk that uses on_process due to the quantity of them. For now at least

File: code/datums/traits/_quirk.dm [hunk 2]
```diff
@@ -10,6 +10,7 @@
...
 	var/mob/living/quirk_holder
 
 /datum/quirk/New(mob/living/quirk_mob, spawn_effects)
```
Comment (ike709) [inline code/datums/traits/_quirk.dm:16]: Might be smart to put a `SHOULD_CALL_PARENT(TRUE)` in here considering how important it is.

File: code/datums/traits/negative.dm [hunk 3]
```diff
@@ -485,14 +490,15 @@
...
+	process = TRUE
+	var/static/drug_list = list(/datum/reagent/drug/crank, /datum/reagent/drug/krokodil, /datum/reagent/medicine/morphine, /datum/reagent/drug/happiness, /datum/reagent/drug/methamphetamine, /datum/reagent/drug/ketamine) //List of possible IDs
 	var/datum/reagent/reagent_type //!If this is defined, reagent_id will be unused and the defined reagent type will be instead.
 	var/datum/reagent/reagent_instance //! actual instanced version of the reagent
 	var/where_drug //! Where the drug spawned
 	var/obj/item/drug_container_type //! If this is defined before pill generation, pill generation will be skipped. This is the type of the pill bottle.
 	var/where_accessory //! where the accessory spawned
 	var/obj/item/accessory_type //! If this is null, an accessory won't be spawned.
-	var/process_interval = 30 SECONDS //! how frequently the quirk processes
+	var/static/process_interval = 30 SECONDS //! how frequently the quirk processes
```
Comment (ike709) [inline code/datums/traits/negative.dm]:
```suggestion
	var/process_interval = 30 SECONDS //! how frequently the quirk processes
```
Unmodified variables are not stored in memory.

File: code/datums/traits/negative.dm [hunk 4]
```diff
@@ -485,14 +490,15 @@
 	gain_text = "<span class='danger'>You suddenly feel the craving for drugs.</span>"
 	lose_text = "<span class='notice'>You feel like you should kick your drug habit.</span>"
 	medical_record_text = "Patient has a history of hard drugs."
-	var/drug_list = list(/datum/reagent/drug/crank, /datum/reagent/drug/krokodil, /datum/reagent/medicine/morphine, /datum/reagent/drug/happiness, /datum/reagent/drug/methamphetamine, /datum/reagent/drug/ketamine) //List of possible IDs
+	process = TRUE
+	var/static/drug_list = list(/datum/reagent/drug/crank, /datum/reagent/drug/krokodil, /datum/reagent/medicine/morphine, /datum/reagent/drug/happiness, /datum/reagent/drug/methamphetamine, /datum/reagent/drug/ketamine) //List of possible IDs
```
Comment (ike709) [inline code/datums/traits/negative.dm]: I don't _think_ this `static` is needed but I could be wrong.

**Review by ike709 — CHANGES_REQUESTED** [review]
See comments.

Didn't check whether `process` is set correctly on each quirk.

Great PR overall.

File: code/datums/traits/negative.dm (hunk 3 above)
Comment (utilbird) [reply code/datums/traits/negative.dm]: _TECHNICALLY_ they are, it's just stored in the data segment ;)

I know what you're saying though

File: code/datums/traits/_quirk.dm (hunk 2 above)
Comment (utilbird) [reply code/datums/traits/_quirk.dm:16]: should_call_parent is only really used on proc definitions. I can add it to this if you want but it seems pretty moot given the thousands of much more important parent procs that don't use it 😆

File: code/datums/traits/negative.dm (hunk 4 above)
Comment (utilbird) [reply code/datums/traits/negative.dm]: Eh, you're probably right, this is somewhat bordering on a net loss given the relative rarity of this datum and frequency of the list access. Not that memory is as much of a problem anymore but still

**Review by ike709 — DISMISSED** [review]

File: code/datums/traits/_quirk.dm (hunk 1 above)
Comment (Fikou) [inline code/datums/traits/_quirk.dm]:
```suggestion
	var/process = FALSE // Does this quirk use on_process()?
```

File: code/controllers/subsystem/processing/quirks.dm [hunk 5]
```diff
@@ -24,8 +24,7 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
 	// Sort by Positive, Negative, Neutral; and then by name
 	var/list/quirk_list = sortList(subtypesof(/datum/quirk), /proc/cmp_quirk_asc)
 
-	for(var/V in quirk_list)
-		var/datum/quirk/T = V
+	for(var/datum/quirk/T as() in quirk_list)
```
Comment (Fikou) [inline code/controllers/subsystem/processing/quirks.dm:27]: as anything is more readable and understandable for coders in the future

**Review by ike709 — APPROVED** [review]

**Pull request merged**